runtime/domain.c

```diff
@@ -648,28 +717,147 @@ CAMLexport void caml_reset_domain_lock(void)
+      ));
+    /* Note: interruptor.running does not guarantee that dom->state is
+       correctly initialized, but domain initialization cannot run
+       concurrently with STW sections so we cannot observe partial
+       initialization states. */
+
+    /* uninitialize the minor heap area */
+    dom->minor_heap_area_start = dom->minor_heap_area_end = 0;
+  }
+
+  size = caml_minor_heaps_end - caml_minor_heaps_start;
+  CAMLassert (Bsize_wsize(caml_minor_heap_max_wsz) * Max_domains == size);
+  caml_mem_unmap((void *) caml_minor_heaps_start, size);
+}
 
+static void stw_resize_minor_heap_reservation(caml_domain_state* domain,
+                                       void* minor_wsz_data,
+                                       int participating_count,
+                                       caml_domain_state** participating) {
+  barrier_status b;
+  uintnat new_minor_wsz = (uintnat) minor_wsz_data;
+
+  caml_gc_log("stw_resize_minor_heap_reservation: "
+              "caml_empty_minor_heap_no_major_slice_from_stw");
+  caml_empty_minor_heap_no_major_slice_from_stw(domain, NULL,
+                                            participating_count, participating);
+
+  caml_gc_log("stw_resize_minor_heap_reservation: free_minor_heap");
+  free_minor_heap();
+
+  b = caml_global_barrier_begin ();
+  if (caml_global_barrier_is_final(b)) {
+    CAML_EV_BEGIN(EV_DOMAIN_RESIZE_HEAP_RESERVATION);
+    caml_gc_log("stw_resize_minor_heap_reservation: "
+                "unreserve_minor_heaps");
+
+    unreserve_minor_heaps();
+    /* new_minor_wsz is page-aligned because caml_norm_minor_heap_size has
+       been called to normalize it earlier.
+    */
+    caml_minor_heap_max_wsz = new_minor_wsz;
+    caml_gc_log("stw_resize_minor_heap_reservation: reserve_minor_heaps");
+    reserve_minor_heaps();
+    /* The call to [reserve_minor_heaps] makes a new reservation,
+       and it also updates the reservation boundaries of each domain
+       by mutating its [minor_heap_area_start{,_end}] variables.
+
+       Thse variables are synchronized by the fact that we are inside
+       a STW section: no other domains are running in parallel, and
+       the participating domains will synchronize with this write by
+       exiting the barrier, before they read those variables in
+       [allocate_minor_heap] below. */
+    CAML_EV_END(EV_DOMAIN_RESIZE_HEAP_RESERVATION);
+  }
+  caml_global_barrier_end(b);
+
+  caml_gc_log("stw_resize_minor_heap_reservation: "
+              "allocate_minor_heap");
+  /* Note: each domain allocates its own minor heap. This seems
+     important to get good NUMA behavior. We don't want a single
+     domain to allocate all minor heaps, which could create locality
+     issues we don't understand very well. */
+  if (allocate_minor_heap(Caml_state->minor_heap_wsz) < 0) {
+    caml_fatal_error("Fatal error: No memory for minor heap");
+  }
+
+  caml_reset_minor_tables(Caml_state->minor_tables);
```

Comment:
Why are we calling `caml_reset_minor_tables` here? I don't remember why we added this and I can't convince myself that it is actually necessary -- otherwise we could add a comment to explain why.

Reply:
In trunk, `reset_minor_tables` is called after the reallocation of the minor heap, at the end of  `caml_set_minor_heap_size`
https://github.com/ocaml/ocaml/blob/7cc6fdf2d872fc0354c2b4b94a5f822666ac7cbc/runtime/minor_gc.c#L122-L135

My impression was that we forgot to add that, due to oversight. Unless there is a good reason why we do not need to reset the minor tables after we free, re-reserve, and allocate all the minor heaps.

Reply:
> The minor tables should be empty at this point -- they get cleared by caml_empty_minor_heap_domain_clear which is called by caml_stw_empty_minor_heap_no_major_slice (both functions live in minor_gc.c).

Ok, you are right, these tables are already cleared and we do not need to call `reset_minor_tables`. I'm removing the call.

Reply:
Thanks! Marking this resolved.